pinot-core/src/main/java/org/apache/pinot/core/udf/UdfExampleBuilder.java

```diff
@@ -0,0 +1,291 @@
+    public EndomorphismNumericTypesBuilder addExampleWithoutNull(String name, Number... values) {
+      Preconditions.checkArgument(values.length == _arity + 2,
+          "Expected %s values, but got %s", _arity + 2, values.length);
+      // Copy the first _arity values and add the value returned when null handling is disabled.
+      Number[] nullableValues = Arrays.copyOf(values, _arity + 1);
+      UdfExample example = UdfExample.create(name, (Object[]) nullableValues)
+          // Ad the expected result when null handling is disabled. This value is at the end of the values array.
+          // and values.length - 1 == _arity + 1.
+          .withoutNull(values[_arity + 1]);
+
+      return addExample(example);
+    }
+
+    /// Like [SingleBuilder#addExample], but for endomorphism functions.
+    public EndomorphismNumericTypesBuilder addExample(UdfExample example) {
+      Preconditions.checkArgument(example.getInputValues().size() == _arity,
+          "Expected %s input values for endomorphism function with arity %s, but got %s",
+          _arity, _arity, example.getInputValues().size());
+      _cases.add(example);
+      return this;
+    }
+
+    public MultiBuilder and(UdfExampleBuilder other) {
+      MultiBuilder multiBuilder = new MultiBuilder();
+      multiBuilder.and(build());
+      multiBuilder.and(other);
+      return multiBuilder;
+    }
+
+    /// Builds the UdfExampleBuilder that generates the examples for endomorphism functions.
+    public UdfExampleBuilder build() {
+      return () -> {
+        Set<FieldSpec.DataType> numericTypes = Set.of(
+            FieldSpec.DataType.INT,
+            FieldSpec.DataType.LONG,
+            FieldSpec.DataType.FLOAT,
+            FieldSpec.DataType.DOUBLE,
+            FieldSpec.DataType.BIG_DECIMAL
+        );
+        Map<UdfSignature, Set<UdfExample>> cases = Maps.newHashMapWithExpectedSize(numericTypes.size());
+        for (FieldSpec.DataType type : numericTypes) {
+          UdfParameter[] paramsAndResult = new UdfParameter[_arity + 1];
+          for (int i = 0; i < _arity; i++) {
+            paramsAndResult[i] = UdfParameter.of("arg" + i, type);
+          }
+          paramsAndResult[_arity] = UdfParameter.result(type);
+          UdfSignature signature = UdfSignature.of(paramsAndResult);
+          Set<UdfExample> newCases = _cases.stream()
+              .map(testCase -> new EndomorphismNumericTypesBuilder.NumericPinotUdfExample(testCase, type))
+              .collect(Collectors.toSet());
+          cases.put(signature, newCases);
+        }
+        return cases;
+      };
+    }
+
+    private static class NumericPinotUdfExample extends UdfExample {
+      private final UdfExample _base;
+      private final FieldSpec.DataType _type;
+
+      public NumericPinotUdfExample(UdfExample base, FieldSpec.DataType type) {
+        Preconditions.checkArgument(base.getResult(true) instanceof Number
```

Comment:
nit: Feel free to ignore :)  
getResult(boolean) is hard to decode by itself. IJ helps by showing param name though. Using aliases for true/false could help as well, e.g. 


```java
public static final boolean NULLABLE = true;
public static final boolean NON_NULL = false;

...

    Preconditions.checkArgument(base.getResult(NULLABLE) instanceof Number || base.getResult(NULLABLE) == null

```

Reply:
That looks like a C code smell. I think it would more java-like to create a getNullableResult() and getNotNullableResult()